File: scala/acronym/src/main/scala/acronym.scala
class Acronym(phrase: String) {

  lazy val abbreviate: String = {
    val a = Acronym.SplitPattern
      .split(phrase)
      .filter(!_.isEmpty)
      // .map(_.head)
      // .mkString
      // .toUpperCase
      println(a.toList)
      a.mkString
  }

}

object Acronym {

  val SplitPattern = """(?=[A-Z][A-Z]?)|[ -]""".r

  def apply(phrase: String) = new Acronym(phrase)

}

File: java/acronym/src/main/java/Acronym.java
import java.util.function.Function;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class Acronym {

  private static final Pattern SPLIT_PATTERN = Pattern.compile("(\\W+|[a-z](?=[A-Z]))");

  public static String generate(String phrase) {
    return SPLIT_PATTERN
      .splitAsStream(phrase)
      .map(Acronym::extractSignificantLetter)
      .collect(Collectors.joining(""));
  }

  private static final String extractSignificantLetter(String word) {
    return word.toUpperCase().substring(0, 1);
  }

}
